ipaserver/plugins/schema.py

```diff
@@ -839,7 +839,7 @@ def execute(self, *args, **kwargs):
         langs = "".join(getattr(context, "languages", []))
 
         if getattr(self.api, "_schema", None) is None:
-            setattr(self.api, "_schema", {})
+            object.__setattr__(self.api, "_schema", {})
```

Comment:
Any reason for using `object.__setattr__` instead of `setattr` ? `self.api` is an instance of `ipalib.API` which inherits from `ReadOnly`, it means that the method  `ReadOnly.__setattr__` is skipped if I understand correctly (the method checks if the attribute is locked).

Reply:
Xmlrpc tests are run in `development` mode, while API is locked only for this mode(for performance reasons). So, no chance to pass as it was.
`object.__setattr__` is a legitimate way to say API: `I know what I do`.
`api._schema` is a cache, it cannot be read-only.

Reply:
Thanks for the explanation, makes sense.

Reply:
Exactly this. The whole ReadOnly bit was to prevent developers from being too clever by half and twiddling with things in the API in a running server. We left the setattr method alone since sometimes it is necessary and stands out like a sore thumb so its **obvious** when we are twiddling things :-)

We used to do this enforcement in "production" mode too but the startup cost is too high.